### DISCUSSION TITLE
git lfs track takes 22+ seconds on any sized repo, with any number of files...

### DISCUSSION THREAD

**Discussion by rafael-f** [git-lfs/git-lfs#5254] [Q&A]
git lfs track is taking a massive amount of time, a few colleagues also reported this issue, this happens on all our repositories that have lfs active.

Is there any way I can debug this? Maybe a verbose way of executing git lfs track?

It's seriously impacting our productivity...

edit: BTW, using on windows 10, git version 2.39.0.windows.2, git-lfs/3.3.0 (GitHub; windows amd64; go 1.19.3; git 77deabdf)

**Comment by bk2204** [comment]
Hey,

Sorry to hear you're having trouble here.  How large is your `.gitattributes` file?  If it's large and you have many items, that might cause `git lfs track` to be slow.

To debug, you could _try_ running with `GIT_TRACE=1` in the environment, but I don't think there's a lot of useful tracing there.  It's worth a try, though.

It might also be that you're running some sort of monitoring software or a third-party antivirus, which can sometimes interfere with Go-based programs like Git LFS.

**Reply by rafael-f** [reply]
This is my gitattributes file:

```
*.dll filter=lfs diff=lfs merge=lfs -text
*.gif filter=lfs diff=lfs merge=lfs -text
*.png filter=lfs diff=lfs merge=lfs -text
*.jpg filter=lfs diff=lfs merge=lfs -text
*.otf filter=lfs diff=lfs merge=lfs -text
*.eot filter=lfs diff=lfs merge=lfs -text
*.ttf filter=lfs diff=lfs merge=lfs -text
*.woff filter=lfs diff=lfs merge=lfs -text
*.woff2 filter=lfs diff=lfs merge=lfs -text
*.exe filter=lfs diff=lfs merge=lfs -text
```

I also think it's something else... It's so annoying that everyone else would complain if it was an git lfs issue :)

If only there was a way I could know what would be affecting it...

GIT_TRACE didn't appear to change anything...

**Reply by bk2204** [reply]
Yeah, Git LFS should not be slow on that file.

Normally, if you set `GIT_TRACE=1` in the environment, it won't change the performance characteristics, but it should print a large amount of debugging output to standard error.  That should show you if there's some part of the process that appears particularly slow, but as you said it seems unlikely that Git LFS is the problem if this isn't a problem for other people.

Are you using any sort of third-party antivirus (that is, one that isn't Windows Defender), and if so, which one is it?  We've seen problems with Avast and Kaspersky in the past, at least.  Sometimes if you're working on a corporate machine, there can be monitoring software which makes some machines slow.  While you might not be able to remove it, it would at least help to know what it is so we can keep track of known problem cases in case other people run into it.

If you _can_ remove it, I'd recommend removing such software and just using Windows Defender, since that is the program least likely to cause problems with Git and Git LFS, and it's the antivirus recommended by most security experts.

Also, it may be helpful to try running the operation under the Windows Subsystem for Linux, which may avoid the kinds of interception that are a problem here.  It's not available in all environments, but if you can use it, it would at least be a helpful information point for debugging.

**Reply by rafael-f** [reply]
Yeah I totally got the TRACE wrong the first time I tried, I forgot to export it before...

Anyway now it actually did show the information, we can see the 20 seconds it takes on the NewLsFiles line, it's either that or the line above...

```
$ git lfs track
09:23:12.039257 exec-cmd.c:237          trace: resolved executable dir: C:/Users/username/AppData/Local/Atlassian/SourceTree/git_local/mingw32/bin
09:23:12.041633 git.c:748               trace: exec: git-lfs track
09:23:12.042639 run-command.c:654       trace: run_command: git-lfs track
09:23:12.081562 trace git-lfs: exec: git 'version'
09:23:12.148440 trace git-lfs: exec: git '-c' 'filter.lfs.smudge=' '-c' 'filter.lfs.clean=' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'rev-parse' '--git-dir' '--show-toplevel'
09:23:12.219503 trace git-lfs: exec: uname
09:23:12.313997 trace git-lfs: exec: cygpath '-w' '.git'
09:23:12.950868 trace git-lfs: exec: cygpath '-w' 'C:/Files/Projects/company/projectname/projectname'
09:23:13.010134 trace git-lfs: exec: git 'config' '--includes' '--local' 'lfs.repositoryformatversion'
09:23:13.089051 trace git-lfs: exec: cygpath '-w' 'C:\Files\Projects\company\projectname\projectname'
09:23:13.149798 trace git-lfs: exec: git 'config' '--includes' '-l'
09:23:13.211902 trace git-lfs: exec: git 'rev-parse' '--is-bare-repository'
09:23:13.274420 trace git-lfs: exec: git 'config' '--includes' '-l' '--blob' ':.lfsconfig'
09:23:13.335329 trace git-lfs: exec: git 'config' '--includes' '-l' '--blob' 'HEAD:.lfsconfig'
09:23:13.397197 trace git-lfs: Install hook: pre-push, force=false, path=C:\Files\Projects\company\projectname\projectname\.git\hooks\pre-push, upgrading...
09:23:13.398506 trace git-lfs: Install hook: post-checkout, force=false, path=C:\Files\Projects\company\projectname\projectname\.git\hooks\post-checkout, upgrading...
09:23:13.399624 trace git-lfs: Install hook: post-commit, force=false, path=C:\Files\Projects\company\projectname\projectname\.git\hooks\post-commit, upgrading...
09:23:13.400818 trace git-lfs: Install hook: post-merge, force=false, path=C:\Files\Projects\company\projectname\projectname\.git\hooks\post-merge, upgrading...
09:23:34.466943 trace git-lfs: NewLsFiles: running in C:\Files\Projects\company\projectname\projectname git ls-files -z --cached --exclude-standard --others
09:23:34.467556 trace git-lfs: exec: git '-c' 'filter.lfs.smudge=' '-c' 'filter.lfs.clean=' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'ls-files' '-z' '--cached' '--exclude-standard' '--others'
09:23:34.542154 trace git-lfs: findAttributeFiles: located .gitattributes
Listing tracked patterns
    *.dll (.gitattributes)
    *.gif (.gitattributes)
    *.png (.gitattributes)
    *.jpg (.gitattributes)
    *.otf (.gitattributes)
    *.eot (.gitattributes)
    *.ttf (.gitattributes)
    *.woff (.gitattributes)
    *.woff2 (.gitattributes)
    *.exe (.gitattributes)
Listing excluded patterns
09:23:34.553759 trace git-lfs: filepathfilter: creating pattern ".git" of type gitignore
09:23:34.554261 trace git-lfs: filepathfilter: creating pattern "**/.git" of type gitignore
09:23:34.555759 trace git-lfs: filepathfilter: accepting "tmp"
09:23:34.862231 exec-cmd.c:237          trace: resolved executable dir: C:/Users/username/AppData/Local/Atlassian/SourceTree/git_local/mingw32/bin
09:23:34.866152 git.c:459               trace: built-in: git config --bool core.sparseCheckout
```



Edit:
Weird thing is that if I run git ls-files -z --cached --exclude-standard --others alone, it executes instantly... I wonder if there's something between the Install hook part and this...

**Reply by bk2204** [reply]
The only thing it should do in between those two things is read the system and the global (per-user) `gitattributes` files.  It's possible one of those is slow or very large in your case.

**Comment by nfgferreira** [comment]
I have the same issue. I use the nice Sourcetree software, which is taking a long time to stage my files. After investigating, the culprit is a `git fls track` command that the software issues right after you click the Stage button. 
I followed the instructions above and found the time is basically wasted in the line ` trace git-lfs: Install hook: post-merge, force=false, path ... hooks\post-merge, upgrading..` exactly like above.

Is it possible we have a pattern?

Any ideas?

Thanks.

**Reply by bk2204** [reply]
Can you open an issue for this, filling out the issue template, and indicating if you're using any antivirus, firewall, or monitoring software other than what comes with your operating system?
